README.md
# Sippy Cup

## A nano framework for http requests proxied to AWS Lambda

[Set Up a Proxy Resource with the Lambda Proxy Integration](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html#api-gateway-set-up-lambda-proxy-integration-on-proxy-resource)

```python
# lambda_function

from sippycup import SippyCup

app = SippyCup()


@app.route('/hello/<string:name>', methods=['GET', 'POST'])
@app.mimetype('text/plain')
def hello_world(name='World'):
    return 'Hello, {0}!'.format(name)


@app.route('/')
def index():
    return app.request


lambda_handler = app.run

```


CHANGES
Sippy Cup Changelog
===================

Version 0.6.0
-------------
- use SIPPYCUP_SCRIPT_NAME_BASE when generating PATH_INFO
- switch to a docker-compose based development environment
- clean up Makefile
- clean up CI config
- implement SAM in demo CloudFormation template for demo app

Version 0.5.4
-------------
- use case insensitive search for Content-Type header in request event
- added a test to ensure POST requests work as expected

Version 0.5.3
-------------
- fix a bug where a TypeError is always raised when no stage variables are
  defined

Version 0.5.2
-------------
- fixed a bug where QUERY_STRING was initialized as None instead of an empty
  string. This would break things like Flask's request.args.get('something')
  if no query params were in the request

Version 0.5.1
-------------
- added support / workaround for setting multiple cookies

Version 0.5
-----------
- added support for Python 3.6 (Thanks to PJ Santoro! -
  https://github.com/abigperson/)
- use API Gateway's resource path when configuring WSGI environment's
  SCRIPT_NAME and PATH_INFO in order to support nested resources
- hide accountId in demo app
- added CloudFormation template to provision AWS resources required for
  setting up demo app
- support for specifying SCRIPT_NAME so that redirection and URL generation
  work as expected when base path mapping is being used in API Gateway. This
  is done by setting a stage variable named SIPPYCUP_SCRIPT_NAME_BASE. See
  README for additional details.


tests/utils.py
def get_apigr(request=None):
    return {
        'body': None,
        'resource': '/',
        'requestContext': {
            'resourceId': 'x1x2x3',
            'apiId': 'y1y2y3',
            'resourcePath': '/',
            'httpMethod': 'GET',
            'requestId': 'aeb82388-a4e8-11e6-a3ed-b5dcbcb63416',
            'accountId': 'z1z2z3',
            'identity': {
                'apiKey': None,
                'userArn': None,
                'cognitoAuthenticationType': None,
                'accessKey': None,
                'caller': None,
                'userAgent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36',
                'user': None,
                'cognitoIdentityPoolId': None,
                'cognitoIdentityId': None,
                'cognitoAuthenticationProvider': None,
                'sourceIp': '127.0.0.1',
                'accountId': None
            },
            'stage': 'testing'
        },
        'queryStringParameters': None,
        'httpMethod': 'GET',
        'pathParameters': None,
        'headers': {
            'Via': '1.1 a536f7c9dbedc2b462a158901fcd8254.cloudfront.net (CloudFront)',
            'Accept-Language': 'en-US,en;q=0.8',
            'CloudFront-Is-Desktop-Viewer': 'true',
            'CloudFront-Is-SmartTV-Viewer': 'false',
            'CloudFront-Is-Mobile-Viewer': 'false',
            'X-Forwarded-For': '104.56.41.168, 54.240.159.59',
            'CloudFront-Viewer-Country': 'US',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Upgrade-Insecure-Requests': '1',
            'X-Forwarded-Port': '443',
            'Host': 'testing.fogbutter.com',
            'X-Forwarded-Proto': 'https',
            'X-Amz-Cf-Id': '03xj4hdvvnK7SX_cVWC2YOLVzl-H-UOX1FfBFcopVr5RYogDGbD6sw==',
            'CloudFront-Is-Tablet-Viewer': 'false',
            'Cache-Control': 'max-age=0',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36',
            'CloudFront-Forwarded-Proto': 'https',
            'Accept-Encoding': 'gzip, deflate, sdch, br'
        },
        'stageVariables': {
            "ben": "was-here",
            "so_was": "red"
        },
        'path': '/'
    }



tests/test_requests.py
import json
from werkzeug.exceptions import MethodNotAllowed, NotFound
from sippycup import SippyCup

app = SippyCup()


@app.route('/test/hello/<string:name>')
@app.mimetype('text/plain')
def hello_world(name):
    return 'Hello, {0}!'.format(name)


@app.route('/test/json/', methods=['POST'])
def hello_json():
    return app.request


@app.route('/')
def index():
    return 'ohai!'


def test_not_found():
    request = {'path': '/garbonzo', 'httpMethod': 'GET'}
    response = app.run(request, None)
    e = NotFound()
    assert response['headers']['Content-Type'] == 'application/json'
    assert response['statusCode'] == e.code
    assert response['body'] == json.dumps(str(e))


def test_method_not_allowed():
    request = {'path': '/test/json/', 'httpMethod': 'GET'}
    response = app.run(request, None)
    e = MethodNotAllowed()
    assert response['headers']['Content-Type'] == 'application/json'
    assert response['statusCode'] == e.code
    assert response['body'] == json.dumps(str(e))


def test_hello_json():
    request = {'path': '/test/json/', 'httpMethod': 'POST'}
    response = app.run(request, None)
    assert response['headers']['Content-Type'] == 'application/json'
    assert response['statusCode'] == 200
    assert response['body'] == json.dumps(request)


def test_hello_world():
    request = {'path': '/test/hello/sippycup', 'httpMethod': 'GET'}
    response = app.run(request, None)
    assert response['headers']['Content-Type'] == 'text/plain'
    assert response['body'] == 'Hello, sippycup!'


def test_index():
    path = '/'
    request = {'path': path, 'httpMethod': 'GET'}
    response = app.run(request, None)
    assert response['headers']['Content-Type'] == 'application/json'
    assert response['body'] == json.dumps('ohai!')


tests/test_response.py
import json
from datetime import date, datetime
from werkzeug.routing import NotFound, MethodNotAllowed
from sippycup import SippyCup
from utils import get_apigr

app = SippyCup()


@app.route('/hello/<string:name>')
@app.route('/hello/', methods=['GET'])
@app.mimetype('text/plain')
def hello_world(name='World'):
    return 'Hello, {0}!'.format(name)


@app.route('/json/')
def json_data():
    return {
        'ohai': 'there',
        'date': date(2016, 1, 1),
        'datetime': datetime(2016, 1, 1, 0, 0, 0, 0)
    }


@app.route('/redirect/')
def redirect():
    app.response.redirect('/going/places', 301)


def test_redirect():
    r = get_apigr()
    r['path'] = '/redirect/'
    response = app.run(r, None)
    print(response)
    assert response['statusCode'] == 301
    assert response['headers']['Location'] == '/going/places'


def test_return_string_default():
    r = get_apigr()
    r['path'] = '/hello/'
    response = app.run(r, None)

    assert 'statusCode' in response
    assert response['statusCode'] == 200
    assert 'Content-Type' in response['headers']
    assert response['headers']['Content-Type'] == 'text/plain'
    assert 'body' in response
    assert response['body'] == 'Hello, World!'


def test_return_string_param():
    r = get_apigr()
    r['path'] = '/hello/SippyCup'
    response = app.run(r, None)

    assert 'statusCode' in response
    assert response['statusCode'] == 200
    assert 'Content-Type' in response['headers']
    assert response['headers']['Content-Type'] == 'text/plain'
    assert 'body' in response
    assert response['body'] == 'Hello, SippyCup!'


def test_redirect_slash():
    r = get_apigr()
    r['path'] = '/hello'
    response = app.run(r, None)
    assert response['statusCode'] == 302
    assert 'Location' in response['headers']
    assert response['headers']['Location'] == \
        'https://testing.fogbutter.com/testing/hello/'


def test_json():
    r = get_apigr()
    r['path'] = '/json/'
    response = app.run(r, None)
    assert response['headers']['Content-Type'] == 'application/json'
    assert json.loads(response['body']) == {
        'ohai': 'there',
        'date': '2016-01-01',
        'datetime': '2016-01-01T00:00:00'
    }


def test_404():
    r = get_apigr()
    r['path'] = '/garbonzo/'
    response = app.run(r, None)
    assert response['statusCode'] == NotFound.code
    assert response['headers']['Content-Type'] == 'text/plain'
    assert response['body'] == NotFound.description


def test_405():
    r = get_apigr()
    r['path'] = '/hello/'
    r['httpMethod'] = 'POST'
    response = app.run(r, None)
    assert response['statusCode'] == MethodNotAllowed.code
    assert response['headers']['Content-Type'] == 'text/plain'
    assert response['body'] == MethodNotAllowed.description


lambda_function.py
# Demo App for Sippy Cup

from flask import Flask, Response, request, jsonify, make_response
from sippycup import sippycup

app = Flask(__name__)

@app.route('/hello/', methods=['GET', 'POST'])
@app.route('/hello/<string:name>', methods=['GET', 'POST'])
def hello_world(name='World'):
    return Response('Hello, {0}!'.format(name), mimetype='text/plain')


@app.route('/')
def index():
    # return the additional WSGI environment variables that SippyCup
    # provides
    rc = request.environ['apigateway.requestContext']
    rc['accountId'] = 'xxxxxxxxxx'
    r = make_response(jsonify({
        'requestContext': rc,
        'stageVariables': request.environ['apigateway.stageVariables']
    }))
    r.set_cookie('cisfor', 'cookie')
    r.set_cookie('cookiestartswith', 'c')
    r.set_cookie('multiplecookies', 'is possible')
    return r


def lambda_handler(event, context):
    return sippycup(app, event, context)


if __name__ == '__main__':
    app.run()


sippycup/responses.py
import json
from datetime import date, datetime
from werkzeug.wrappers import Response


class SippyCupResponse(Response):

    @property
    def apigr(self):
        return {
            'statusCode': self.status_code,
            'body': self.get_data(True),
            'headers': {key: value for key, value in self.headers.iteritems()}
        }

    def __init__(self, response=None, status=None, headers=None,
                 mimetype=None, content_type=None, direct_passthrough=False):
        super(SippyCupResponse, self).__init__(
            response=response, status=status, headers=headers,
            mimetype=mimetype, content_type=content_type,
            direct_passthrough=direct_passthrough
        )
        self.headers['Content-Type'] = 'application/json'

    def redirect(self, url, code=302):
        self.status_code = code
        self.headers['Location'] = url

    def set_data(self, value):
        if self.headers['Content-Type'] == 'application/json':
            value = json.dumps(value, default=self.json_serializer)

        return super(SippyCupResponse, self).set_data(value)

    @staticmethod
    def json_serializer(obj):
        if type(obj) in [datetime, date]:
            return obj.isoformat()
        else:
            raise TypeError('{0}.{1} is not JSON serializable'.format(
                type(obj).__module__, type(obj).__name__
            ))


sippycup/wsgienviron.py
from StringIO import StringIO
from urllib import urlencode

# https://www.python.org/dev/peps/pep-0333/


class WsgiEnviron(object):

    @property
    def body(self):
        if self.request['body'] is not None:
            return StringIO(self.request['body'])
        else:
            return None

    @property
    def content_length(self):
        if self.request['body'] is not None:
            return str(len(self.request['body']))
        else:
            return str(0)

    @property
    def content_type(self):
        try:
            return self.request['headers']['Content-Type']
        except KeyError:
            return None

    @property
    def url_scheme(self):
        if 'X-Forwarded-Proto' in self.request['headers']:
            return self.request['headers']['X-Forwarded-Proto']
        else:
            return 'http'

    @property
    def server_port(self):
        if self.url_scheme == 'https':
            return '443'
        else:
            return '80'

    @property
    def query_string(self):
        if self.request['queryStringParameters'] is None:
            return None
        else:
            return urlencode(self.request['queryStringParameters'])

    def __init__(self, request):
        self.request = request

        self.environ = {
            'REQUEST_METHOD': request['httpMethod'],
            'SCRIPT_NAME': '/{0}'.format(request['requestContext']['stage']),
            'PATH_INFO': request['path'],
            'QUERY_STRING':  self.query_string,
            'CONTENT_TYPE': self.content_type,
            'CONTENT_LENGTH': self.content_length,
            'SERVER_NAME':  request['headers']['Host'],
            'SERVER_PORT': self.server_port,
            'SERVER_PROTOCOL': 'HTTP/1.1',
            'wsgi.version': (1, 0),
            'wsgi.url_scheme': self.url_scheme,
            'wsgi.input': self.body,
            'wsgi.errors': '',  # what should this be?
            'wsgi.multiprocess': False,
            'wsgi.multithread': False,
            'wsgi.run_once': False
        }

        headers = {'HTTP_{0}'.format(key.upper().replace('-', '_')): value
                   for key, value in request['headers'].iteritems()}

        self.environ.update(headers)


sippycup/__init__.py
import itertools
from sys import stderr, stdout
try:
    from urllib import urlencode
    from StringIO import StringIO
except ImportError:
    from urllib.parse import urlencode
    from io import StringIO


def sippycup(app, event, context=None):
    response = SippyCupResponse()
    result = app(WsgiEnviron(event).environ, response)
    return response.apigr(result)


class SippyCupResponse:

    def apigr(self, body=''):

        # The response format states that headers must be a dictionary, which
        # presents a problem when multiple Set-Cookie headers are returned by
        # the application (only the last cookie set gets put into the response)
        #
        # Browser's seem not to care about the case of the header name, and
        # API Gateway does not alter the header names.
        #
        # As such, setting multiple cookies in the response seems to be
        # possible by simply using a different case permutations of 'Set-Cookie'
        #
        # Generation of case permutations comes from ephemient on StackOverflow
        # https://stackoverflow.com/questions/11144389/find-all-upper-lower-and-mixed-case-combinations-of-a-string

        headers = {h[0]: h[1] for h in self.headers
                   if h[0].lower() != 'set-cookie'}

        headers.update(zip(
            map(
                ''.join,
                itertools.product(
                    *zip('set-cookie'.upper(), 'set-cookie'.lower()))
            ),
            [h[1] for h in self.headers if h[0].lower() == 'set-cookie']
        ))

        return {
            'statusCode': int(self.status.split()[0]),
            'body': ''.join(
                [self.body.getvalue()] +
                [x.decode(stdout.encoding or 'UTF8') for x in body]
            ),
            'headers': headers
        }

    def __init__(self):
        self.status = '200 OK'
        self.headers = []
        self.body = StringIO()

    def __call__(self, status, headers, exc_info=None):
        self.status = str(status)
        self.headers = list(headers)
        return self.body.write


class WsgiEnviron(object):

    # https://www.python.org/dev/peps/pep-0333/

    @property
    def body(self):
        if self.request['body'] is not None:
            return StringIO(self.request['body'])
        else:
            return None

    @property
    def content_length(self):
        if self.request['body'] is not None:
            return str(len(self.request['body']))
        else:
            return str(0)

    @property
    def content_type(self):
        header = [val for key, val in self.request['headers'].items()
                  if key.lower() == 'content-type']
        try:
            return header[0]
        except IndexError:
            return None

    @property
    def query_string(self):
        if self.request['queryStringParameters'] is None:
            return ''
        else:
            return urlencode(self.request['queryStringParameters'])

    @property
    def script_name(self):
        if 'SIPPYCUP_SCRIPT_NAME_BASE' in self.request['stageVariables']:
            base = self.request['stageVariables']['SIPPYCUP_SCRIPT_NAME_BASE'].rstrip('/')
        else:
            base = ''.join(['/', self.request['requestContext']['stage']])

        return '/'.join([base, self.resource_path_base.lstrip('/')]).rstrip('/')

    @property
    def resource_path_base(self):
        end = self.request['requestContext']['resourcePath'].rfind('/{proxy+}')
        if end == -1:
            end = None
        return self.request['requestContext']['resourcePath'][0:end].rstrip('/')

    @property
    def path_info(self):
        if 'SIPPYCUP_SCRIPT_NAME_BASE'in self.request['stageVariables']:
            return self.request['path'].replace(self.script_name, '', 1)
        else:
            return self.request['path'].replace(self.resource_path_base, '', 1)

    def __init__(self, request):

        if request['stageVariables'] is None:
            request['stageVariables'] = {}

        self.request = request

        self.environ = {
            'REQUEST_METHOD': request['httpMethod'],
            'SCRIPT_NAME': self.script_name,
            'PATH_INFO': self.path_info,
            'QUERY_STRING':  self.query_string,
            'CONTENT_TYPE': self.content_type,
            'CONTENT_LENGTH': self.content_length,
            'SERVER_NAME':  request['headers']['Host'],
            'SERVER_PORT': '443',
            'SERVER_PROTOCOL': 'HTTP/1.1',
            'wsgi.version': (1, 0),
            'wsgi.url_scheme': 'https',
            'wsgi.input': self.body,
            'wsgi.errors': stderr,
            'wsgi.multiprocess': False,
            'wsgi.multithread': False,
            'wsgi.run_once': False,
            'apigateway.stageVariables': request['stageVariables'],
            'apigateway.requestContext': request['requestContext'],
            'apigateway': True
        }

        headers = {'HTTP_{0}'.format(key.upper().replace('-', '_')): value
                   for key, value in request['headers'].items()}

        self.environ.update(headers)
